pandas/core/groupby.py

```diff
@@ -3742,7 +3742,22 @@ def _reindex_output(self, result):
         levels_list = [ping.group_index for ping in groupings]
         index = MultiIndex.from_product(levels_list, names=self.grouper.names)
         d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
-        return result.reindex(**d).sortlevel(axis=self.axis)
+
+        if not self.as_index:
+            result.set_index(self.grouper.result_index, inplace=True)
+
+        result = result.reindex(**d).sortlevel(axis=self.axis)
+
+        if not self.as_index:
+            idx = result.index
```

Comment:
this is pretty kludgey - maybe an do this inside the Index (not exactly sure what you are trying to do here)